### ISSUE TITLE
Calibration generates negative values

### ISSUE THREAD

**Issue by eliebnn** [rust-dd/stochastic-rs#14]
Hi,

I have tried to run the calibration code. Some runs return negative values for v0, theta, kappa, signa, which should all be strictly positive.

Attached is a code sample. It runs the calibration over several v0, and keep the minimul mape loss. 

Also, if you uncomment the `set_initial_params` line, you will see that sigma and rho will have NaN assigned values.

Am I doing something wrong? I run the calibration over call options with 1 months maturity with same spot, and different strikes.

```
use stochastic_rs::quant::calibration::heston::{HestonCalibrator, HestonParams};
use stochastic_rs::quant::OptionType;

use std::cmp::Ordering;
use ndarray::Array1;

#[tokio::main]
async fn main() {

    let r = 0.04;
    let q = 0.06;
    let t = 0.083;

    let strikes_ls: Vec<f64> = vec![5220.318, 6090.371, 6960.424, 7830.477, 8265.5035, 8483.01675, 8700.53, 8918.04325, 9135.5565, 9570.583, 10440.636, 11310.689];
    let spots_ls: Vec<f64> = vec![8700.53, 8700.53, 8700.53, 8700.53, 8700.53, 8700.53, 8700.53, 8700.53, 8700.53, 8700.53, 8700.53, 8700.53];
    let vol_ls: Vec<f64> = vec![0.3669, 0.3082, 0.2218, 0.1799, 0.1393, 0.1156, 0.1019, 0.0923, 0.0915, 0.1086, 0.1237, 0.136];
    let markets_ls: Vec<f64> = vec![3499.564, 2632.74, 1765.933, 901.846, 479.055, 279.313, 118.848, 28.79, 4.23, 0.143, 1.799e-5, 1.259e-9];

    let v0 = Array1::linspace(1e-5, 2.0, 5);

    for v in v0.iter() {
      let mut calibrator = HestonCalibrator::new(
        HestonParams {
            v0: *v,
            theta: 6.47e-5,
            rho: -1.98e-3,
            kappa: 6.57e-3,
            sigma: 5.09e-4,
        },
        markets_ls.clone().into(), spots_ls.clone().into(), strikes_ls.clone().into(),
        t, r, Some(q), OptionType::Call,
      );

      // Uncomment to witness NaN Sigma/Rho
      // calibrator.set_initial_params(spots_ls.clone().into(), vol_ls.clone().into(), r);

      let result = calibrator.calibrate();
      let result_unwrap = result.unwrap();

      let minimum_loss = result_unwrap.iter().min_by(|a, b| {
        a.loss_scores.mape.partial_cmp(&b.loss_scores.mape).unwrap_or(Ordering::Equal)
      });

      println!("{:?}", minimum_loss.clone().unwrap().loss_scores.mape);
      println!("{:?}", minimum_loss.clone().unwrap().params);

    }
    }
```

The `markets_ls` array was computed with:

```
    let markets_ls: Vec<f64> = data.iter().map(|x| {

        let bs = BSMPricer::new(x.spot, x.vol, x.strike, r, None, None, Some(q), Some(t), None, None, OptionType::Call, BSMCoc::BSM1973);
        let (c, _) = bs.calculate_call_put();

        return c;

    }).collect()
```

**Comment by dancixx** [comment]
@eliebnn I have fixed that if the initial guess is missing you have to call the `set_initial_params` method, which is a MLE of the Heston params. I have tested your input data and the optimized params are all positive, but for deeper analysis, why is the optimizer not so accurate I need some more time to investigate.

Also, I think that the NMLE of the Heston params is not the most accurate.

**Comment by eliebnn** [comment]
@dancixx 

Let me be more precise - as you can see there is a commented line running `calibrator.set_initial_params`

1. If I *don't* run it, aka I keep it commented and I rely on the params input at HestonCalibrator instantiation level as initial guess, I get:

```
47.288979999331175
HestonParams { v0: 0.035043074836283605, theta: 0.023307280430281397, rho: 0.029265462386466674, kappa: 0.02967394730385698, sigma: 0.03846709839019855 }
66.20941540495458
HestonParams { v0: 0.37227915578672055, theta: -0.15197430375836488, rho: -0.10932777647163476, kappa: -0.12056909932504765, sigma: -0.11646108235334397 }
54.28685598369967
HestonParams { v0: -1.1149357388589973, theta: -2.480726552436385, rho: -1.783973957668885, kappa: -2.1017576315752713, sigma: -1.951665906481375 }
54.24134844899703
HestonParams { v0: -1.2908904357378, theta: -3.0435711593039483, rho: -2.460311091556809, kappa: -2.7477318372936614, sigma: -2.6908963614262706 }
49.40360359951317
HestonParams { v0: -0.4039161192241324, theta: -4.146670889846464, rho: -2.7055853232409834, kappa: -3.507614551413719, sigma: -3.846146800654562 }
```

**Note**: negative values are generated.

2. If I *do* run it, aka uncomment `calibrator.set_initial_params`, I get:

```
77.16060137226997
HestonParams { v0: 0.3669, theta: 0.18611966316586462, rho: NaN, kappa: -5.652905371667904, sigma: NaN }
77.16060137226997
HestonParams { v0: 0.3669, theta: 0.18611966316586462, rho: NaN, kappa: -5.652905371667904, sigma: NaN }
77.16060137226997
HestonParams { v0: 0.3669, theta: 0.18611966316586462, rho: NaN, kappa: -5.652905371667904, sigma: NaN }
77.16060137226997
HestonParams { v0: 0.3669, theta: 0.18611966316586462, rho: NaN, kappa: -5.652905371667904, sigma: NaN }
77.16060137226997
HestonParams { v0: 0.3669, theta: 0.18611966316586462, rho: NaN, kappa: -5.652905371667904, sigma: NaN }
```

**Note**: NaN values are generated.


Are you able to replicate this by running my script?

**Comment by dancixx** [comment]
@eliebnn I will take a look again soon.

**Comment by dancixx** [comment]
@eliebnn I have checked your example and you are right. I guess (after some debugging) we have some calculation error in the NMLE method which is a maximum likelihood estimation for the initial guess based on this paper: [http://scis.scichina.com/en/2018/042202.pdf](http://scis.scichina.com/en/2018/042202.pdf).

The second issue can be in the optimizer Jacobian matrix that is also implemented based on this paper: [https://www.sciencedirect.com/science/article/abs/pii/S0377221717304460](https://www.sciencedirect.com/science/article/abs/pii/S0377221717304460)

These are fairly formula-heavy algorithms. I have to go through the implementation again, where we have a typo or other bugs. Thx for your patience.

**Comment by eliebnn** [comment]
Hi @dancixx, have you had some luck with that issue?

**Comment by eliebnn** [comment]
Any chance you can get ideas from the way the Heston model is implemented in the C++ QuantLib library? Saying this but I'm unsure they use the same methodology altogether.

**Comment by dancixx** [comment]
@eliebnn t seems my assumption was correct — the derivatives were indeed the source of the problem. I have recalculated them, and I am now getting positive values in your example. Please take a look, and thank you for your patience. There is still an error when trying to take an initial guess using maximum likelihood estimation, but I need more time to fix that.

**Comment by eliebnn** [comment]
@dancixx did you get the chance to look into the gradient calculation?

**Comment by dancixx** [comment]
@eliebnn I was to busy last month. I will take a look in the following weeks.

**Comment by dancixx** [comment]
@eliebnn i will add the constraint, btw during the tests I got sometimes negative value but in edge cases but lets see. The loss scores are still there. I will also add a flag to activate this easier.

**Comment by eliebnn** [comment]
The Calibration History attribute is private, so it cannot be explored.

**Comment by eliebnn** [comment]
I'm unsure the new flag `record_history` has any effect at the moment, because the storage actually happens in the HestonCalibrator cloned instance called `problem`.

Hence, this returns an empty array:

```rust
calibrator.calibrate();
let result = calibrator.history();
println!("{:?}", result);
```
